source/postprocess/visualization/density_anomaly.cc

```diff
@@ -0,0 +1,244 @@
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      template <int dim>
+      DensityAnomaly<dim>::
+      DensityAnomaly ()
+        :
+        CellDataVectorCreator<dim>("kg/m^3")
+      {}
+
+      template <int dim>
+      std::pair<std::string, std::unique_ptr<Vector<float>>>
+      DensityAnomaly<dim>::execute() const
+      {
+        std::pair<std::string, std::unique_ptr<Vector<float>>>
+        return_value ("density_anomaly",
+                      std::make_unique<Vector<float>>(this->get_triangulation().n_active_cells()));
+
+        // Calculate the maximum depth of the domain
+        const double max_depth = this->get_geometry_model().maximal_depth();
+
+        // The following lines evaluate the density at a single point per cell
+        // and then compute the difference from the average in
+        // the slice within which the point lies
+        const QMidpoint<dim> quadrature_formula;
+        const unsigned int n_q_points = quadrature_formula.size();
+
+
+        FEValues<dim> fe_values (this->get_mapping(),
+                                 this->get_fe(),
+                                 quadrature_formula,
+                                 update_values   |
+                                 update_gradients |
+                                 update_quadrature_points );
+
+        MaterialModel::MaterialModelInputs<dim> in(n_q_points, this->n_compositional_fields());
+        MaterialModel::MaterialModelOutputs<dim> out(n_q_points, this->n_compositional_fields());
+
+        switch (average_density_scheme)
+          {
+            case reference_profile:
+            {
+              MaterialModel::MaterialModelOutputs<dim> adiabatic_out(n_q_points, this->n_compositional_fields());
+
+              // Loop over the cells
+              for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+                if (cell->is_locally_owned())
+                  {
+                    // Get the pressure, temperature and composition in the cell
+                    fe_values.reinit (cell);
+
+                    in.reinit(fe_values, cell, this->introspection(), this->get_solution());
+
+                    // Evaluate density at the cell center
+                    this->get_material_model().evaluate(in, out);
+                    const double density = out.densities[0];
+
+                    // Substitute the adiabatic reference state for temperature and pressure,
+                    // then reevaluate the material model.
+                    in.temperature[0]=this->get_adiabatic_conditions().temperature(in.position[0]);
+                    in.pressure[0]=this->get_adiabatic_conditions().pressure(in.position[0]);
+
+                    // Evaluate adiabatic/reference density at the cell center
+                    this->get_material_model().evaluate(in, adiabatic_out);
+                    const double adiabatic_density = adiabatic_out.densities[0];
```

Comment:
is it important that you evaluate the density using the adiabatic temperature and adiabatic pressure, instead of just asking for the density of the adiabatic profile?
In other words couldnt you just do:
```suggestion
                    const double adiabatic_density = this->get_adiabatic_conditions().density(in.position[0]);
```

The only difference I see is that the adiabatic density in the profile is evaluated for an assumed reference composition, while your implementation uses the actual current composition distribution. But I think your implementation actually contradicts the description of the postprocessor a bit, because your adiabatic density is actually different in different horizontal positions of the model. Maybe you have a reason for your implementation, but it would be good to document this here.